### PR TITLE
feat(swarm): allow both Disconnected and NotDialing condition combined

### PR DESCRIPTION
**Pull request by b-zee**
## Description

Dialing by default uses the condition `Disconnected`, but it seems appropriate to additionally prevent a dial if already dialing. This means a combination of `Disconnected` and `NotDialing`.

The default is changed to a new variant `DisconnectedAndNotDialing`. This is a breaking change.

Related: #4189.

## Change checklist

- [ ] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by thomaseizinger** [comment]
I moved this to draft to prevent it being merged accidentally. I am surprised cargo semver-checks doesn't pick up on this?

**Comment by b-zee** [comment]
`semver-checks` does pick it up for me:

```
$ cargo semver-checks --baseline-rev origin/master --package=libp2p-swarm
     Cloning origin/master
     Parsing libp2p-swarm v0.43.1 (current)
     Parsing libp2p-swarm v0.43.1 (baseline)
    Checking libp2p-swarm v0.43.1 -> v0.43.1 (no change)
   Completed [   0.057s] 45 checks; 44 passed, 1 failed, 0 unnecessary

--- failure enum_variant_added: enum variant added on exhaustive enum ---

Description:
A publicly-visible enum without #[non_exhaustive] has a new variant.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#enum-variant-new
       impl: https://github.com/obi1kenobi/cargo-semver-check/tree/v0.22.1/src/lints/enum_variant_added.ron

Failed in:
  variant PeerCondition:DisconnectedAndNotDialing in /home/b-zee/src/rust-libp2p/swarm/src/dial_opts.rs:325
       Final [   0.058s] semver requires new major version: 1 major and 0 minor checks failed
```

**Comment by thomaseizinger** [comment]
I noticed that we are two versions behind so maybe that is why :)

**Review by mxinden — COMMENTED** [review]
Thank you. To me this is the most concise solution to the problem.

Would you mind adding a changelog entry to `swarm/CHANGELOG.md` and bump the crate version in `swarm/Cargo.toml`? Don't worry about other version bumps. For now it is fine for CI to be red. I just want to make sure we don't forget the changelog entry and the version bump once we cut v0.53.0.

Thank you for the continued work.

**Comment by b-zee** [comment]
Done and rebased to fix merge conflicts.

File: swarm/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,11 @@
+## 0.44.0 - unreleased
+
+- Add `PeerCondition` variant `DisconnectedAndNotDialing`, combining pre-existing conditions.
+  A new dialing attempt is iniated _only if_ the peer is both considered disconnected and there is currently no ongoing dialing attempt.
+  See [PR 4225].
```
Comment (thomaseizinger) [inline swarm/CHANGELOG.md]: We should mention that this is the new default.

**Comment by mxinden** [comment]
> This pull request has merge conflicts. Could you please resolve them @b-zee? pray

No need to resolve. Conflicts are simple. Let's resolve them close to v0.53 release unless difficult conflicts come up.

**Comment by thomaseizinger** [comment]
> > This pull request has merge conflicts. Could you please resolve them @b-zee? pray
> 
> No need to resolve. Conflicts are simple. Let's resolve them close to v0.53 release unless difficult conflicts come up.

We are thinking of cutting a new release soon, mind resolving the conflicts @b-zee?

**Comment by b-zee** [comment]
Rebased. The changes in master that were supposed to be a minor swarm release of 0.43.4, I have changed to 0.44.0, as this PR requires that if I'm not mistaken.

**Review by thomaseizinger — COMMENTED** [review]
Thanks for the update! There seems to be a failing test still :)

**Comment by b-zee** [comment]
> There seems to be a failing test still :)

I'm not sure how to attack that one. I don't think I changed anything dependency wise.

Anyhow, just rebased again and that might even fix the `cargo-deny` action. 🤞

**Comment by thomaseizinger** [comment]
> > There seems to be a failing test still :)
> 
> I'm not sure how to attack that one. I don't think I changed anything dependency wise.

It seems the only failing one is in `libp2p-memory-connection-limits` which I am guessing is just a missing and now explicitly required `DialCondition` to not use the new default?

> Anyhow, just rebased again

Merging would be preferred! :pray: 

See https://github.com/libp2p/rust-libp2p/blob/master/CONTRIBUTING.md#we-squash-merge-pull-requests.

**Comment by b-zee** [comment]
Pushed the fix for the test. I've taken the liberty to use the `Always` condition. As that test should dial always, but only fail by the connection limit itself. In theory, a connection could have been established by one of the dials, making it fail for that reason and not for what is described in the test.

**Comment by thomaseizinger** [comment]
> Pushed the fix for the test. I've taken the liberty to use the `Always` condition. As that test should dial always, but only fail by the connection limit itself. In theory, a connection could have been established by one of the dials, making it fail for that reason and not for what is described in the test.

Sounds good! Can you apply the same fix to the other tests?

**Comment by b-zee** [comment]
Now that I've come back from vacation I have looked at it proper. Tests are fixed now, sorry for the back and forth.

The `mdns` test fails. I've looked at a bit, but I do not see the relation between that test and my changes. I'm quite sure the dialing changes some behavior here, but I can't pick up on it. Do you have any thoughts, @thomaseizinger?

**Comment by thomaseizinger** [comment]
> The `mdns` test fails. I've looked at a bit, but I do not see the relation between that test and my changes. I'm quite sure the dialing changes some behavior here, but I can't pick up on it. Do you have any thoughts, @thomaseizinger?

I've restarted that one. It is probably just a rare flake :/

File: swarm/CHANGELOG.md [hunk 2]
```diff
@@ -1,4 +1,8 @@
-## 0.43.4
+## 0.44.0 - unreleased
+
+- Add `PeerCondition` variant `DisconnectedAndNotDialing`, combining pre-existing conditions. This is the new default.
```
Comment (thomaseizinger) [inline swarm/CHANGELOG.md]:
```suggestion
- Add `PeerCondition::DisconnectedAndNotDialing` variant, combining pre-existing conditions.
  This is the new default.
```

**Review by thomaseizinger — APPROVED** [review]

**Comment by thomaseizinger** [comment]
Setting this to `draft` because it is part of the next milestone.

File: swarm/CHANGELOG.md [hunk 3]
```diff
@@ -36,6 +40,7 @@
 - Introduce `SwarmBuilder::idle_conncetion_timeout` and deprecate `keep_alive::Behaviour` as a result.
   See [PR 4161].
 
+[PR 4225]: https://github.com/libp2p/rust-libp2p/pull/4225
```
Comment (thomaseizinger) [inline swarm/CHANGELOG.md]:
```suggestion
```

File: swarm/CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,9 @@
 ## 0.44.0 - unreleased
 
+- Add `PeerCondition::DisconnectedAndNotDialing` variant, combining pre-existing conditions.
+  This is the new default.
+  A new dialing attempt is iniated _only if_ the peer is both considered disconnected and there is currently no ongoing dialing attempt.
+  See [PR 4225].
```
Comment (thomaseizinger) [inline swarm/CHANGELOG.md]:
```suggestion
  See [PR 4225](https://github.com/libp2p/rust-libp2p/pull/4225).
```

**Comment by b-zee** [comment]
Thanks @thomaseizinger, wasn't sure when this would be merged so didn't want to keep the mergify bot happy :grin:

**Pull request merged**